sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -16,84 +16,42 @@
 
 package com.nvidia.spark.rapids
 
-import ai.rapids.cudf.Scalar
 import com.nvidia.spark.rapids.RapidsPluginImplicits._
 
 import org.apache.spark.sql.catalyst.analysis.{TypeCheckResult, TypeCoercion}
 import org.apache.spark.sql.catalyst.expressions.{ComplexTypeMergingExpression, Expression}
 import org.apache.spark.sql.types.{BooleanType, DataType}
 import org.apache.spark.sql.vectorized.ColumnarBatch
 
-abstract class GpuConditionalExpression extends ComplexTypeMergingExpression with GpuExpression {
+trait GpuConditionalExpression extends ComplexTypeMergingExpression with GpuExpression {
 
   protected def computeIfElse(
       batch: ColumnarBatch,
-      predicateExpr: Expression,
+      predExpr: Expression,
       trueExpr: Expression,
-      falseValues: GpuColumnVector): GpuColumnVector = {
-    withResource(GpuExpressionsUtils.columnarEvalToColumn(predicateExpr, batch)) { predicate =>
-      val trueResult: Any = trueExpr.columnarEval(batch)
-      try {
-        val result = trueResult match {
-          case t: GpuColumnVector => predicate.getBase.ifElse(t.getBase, falseValues.getBase)
-          case t: GpuScalar => predicate.getBase.ifElse(t.getBase, falseValues.getBase)
-          case u =>
-            throw new IllegalStateException(s"Unexpected inputs $u")
-        }
-        GpuColumnVector.from(result, dataType)
-      } finally {
-        trueResult match {
-          case a: AutoCloseable => a.close()
-          case _ =>
-        }
-      }
-    }
-  }
-
-  protected def computeIfElse(
-      batch: ColumnarBatch,
-      predicateExpr: Expression,
-      trueExpr: Expression,
-      falseValue: Scalar): GpuColumnVector = {
-    withResource(GpuExpressionsUtils.columnarEvalToColumn(predicateExpr, batch)) { predicate =>
-      val trueResult: Any = trueExpr.columnarEval(batch)
-      try {
-        val result = trueResult match {
-          case t: GpuColumnVector => predicate.getBase.ifElse(t.getBase, falseValue)
-          case t: GpuScalar => predicate.getBase.ifElse(t.getBase, falseValue)
-          case u =>
-            throw new IllegalStateException(s"Unexpected inputs $u")
-        }
-        GpuColumnVector.from(result, dataType)
-      } finally {
-        trueResult match {
-          case a: AutoCloseable => a.close()
-          case _ =>
+      falseValue: Any): GpuColumnVector = {
+    withResource(GpuExpressionsUtils.columnarEvalToColumn(predExpr, batch)) { pred =>
+      withResourceIfAllowed(trueExpr.columnarEval(batch)) { trueRet =>
+        withResourceIfAllowed(falseValue) { falseRet =>
```

Comment:
Please move this to be the first thing that we do. Otherwise if there is an exception thrown when computing `trueExpr` or `predExpr` the false expression will leak.

Reply:
Nice catch. Updated